### PR TITLE
Including EVAC+ and util function

### PR DESCRIPTION
**Pull request by pjsjongsung**
This introduces EVAC+  https://doi.org/10.48550/arXiv.2206.02837 and util functions in utils.py to be shared between deep learning models. I have also added some messages that encourages mac users to install tensorflow through the special github link instead of pip. (Since there have been some errors where it does not work properly)

### CODE REVIEW

File: dipy/nn/evac.py [hunk 1]
```diff
@@ -0,0 +1,455 @@
...
+    voxsize : tuple
+    ori_shape : tuple
+
+    Returns
+    -------
+    recovered_img : np.ndarray
+    """
+    new_image, _ = reslice(image, np.eye(4), (2, 2, 2), (1, 1, 1))
+    recovered_img = affine_transform(new_image, affine, output_shape=ori_shape)
+    return recovered_img
```
Comment (skoudoro) [inline dipy/nn/evac.py:91]: why all this function are not in `nn.utils`?    I mean `prepare_img`, `recover_img`, `transform_img`, `set_logger_level` ?

Comment (skoudoro) [reply dipy/nn/evac.py:91]: oh,I see , many fixed size...  maybe they can be set as default arguments

File: dipy/nn/evac.py [hunk 2]
```diff
@@ -0,0 +1,455 @@
...
+            temp_pred = self.__predict(temp_input)
+            prediction[:batch_idx] = temp_pred
+        remainder = np.mod(len(T1), batch_size)
+        if remainder != 0:
+            temp_input = prepare_img(input_data[..., -remainder:])
+            temp_pred = self.__predict(temp_input)
+            prediction[-remainder:] = temp_pred
+
+        output_mask = []
+        for i in range(len(T1)):
```
Comment (skoudoro) [inline dipy/nn/evac.py:438]: `enumerate` instead of `range(len(`

File: dipy/nn/evac.py [hunk 3]
```diff
@@ -0,0 +1,455 @@
...
+        affine : np.ndarray (...) or (batch, ...)
+            affine matrix of mask
+            only if return_affine is True
+
+        """
+        
+        voxsize = np.array(voxsize)
+        affine = np.array(affine)
+        
+        if type(T1) is list or type(T1) is tuple or len(T1.shape) == 4:
```
Comment (skoudoro) [inline dipy/nn/evac.py:396]: use `isinstance(T1, (list, tuple))` instead

File: dipy/nn/evac.py [hunk 4]
```diff
@@ -0,0 +1,455 @@
...
+        r"""
+        Wrapper function to facilitate prediction of larger dataset.
+
+        Parameters
+        ----------
+        T1 : np.ndarray or list of np.ndarrys
+            For a single image, input should be a 3D array.
+            If multiple images, it should be a 4D array or a list.
+
+        affine : np.ndarray (4, 4) or (batch, 4, 4)
```
Comment (skoudoro) [inline dipy/nn/evac.py:352]: `, optional`

File: dipy/nn/evac.py [hunk 5]
```diff
@@ -0,0 +1,455 @@
...
+            or list of np.ndarrays with len of batch
+            Affine matrix for the T1 image. Should have
+            batch dimension if T1 has one.
+            Default is None
+        
+        voxsize : np.ndarray or list or tuple
```
Comment (skoudoro) [inline dipy/nn/evac.py:358]: `, optional`

File: dipy/nn/evac.py [hunk 6]
```diff
@@ -0,0 +1,455 @@
...
+            (3,) or (batch, 3)
+            voxel size of the T1 image.
+            Default is (1, 1, 1)
+
+        batch_size : int
```
Comment (skoudoro) [inline dipy/nn/evac.py:363]: `, optional`

File: dipy/nn/evac.py [hunk 7]
```diff
@@ -0,0 +1,455 @@
...
+            Number of images per prediction pass. Only available if data
+            is provided with a batch dimension.
+            Consider lowering it if you get an out of memory error.
+            Increase it if you want it to be faster and have a lot of data.
+            If None, batch_size will be set to 1 if the provided image
+            has a batch dimension.
+            Default is None
+
+        return_affine : bool
```
Comment (skoudoro) [inline dipy/nn/evac.py:372]: `, optional`

File: dipy/nn/evac.py [hunk 8]
```diff
@@ -0,0 +1,455 @@
...
+            Whether to return the affine matrix. Useful if the input was a
+            file path.
+            Default is False
+
+        return_prob : bool
```
Comment (skoudoro) [inline dipy/nn/evac.py:377]: `, optional`

File: dipy/nn/evac.py [hunk 9]
```diff
@@ -0,0 +1,455 @@
...
+            affine = np.expand_dims(affine, 0)
+            voxsize = np.expand_dims(voxsize, 0)
+            
+
+        input_data = np.zeros((128, 128, 128, len(T1)))
+        rev_affine = np.zeros((len(T1), 4, 4))
+
+        # Normalize the data.
+        n_T1 = np.zeros(T1.shape)
+        for i in range(len(T1)):
```
Comment (skoudoro) [inline dipy/nn/evac.py:416]: `enumerate` instead `range(len`. or do `for t in T1`

**Review by skoudoro — COMMENTED** [review]
Hi @pjsjongsung,

Here a quick initial review.  I need to try it

**Review by skoudoro — COMMENTED** [review]
Thank you for the quick update @pjsjongsung!

Can you also add test for test_utils new functions ?

**Comment by pjsjongsung** [comment]
@skoudoro test_utils should be here! If it looks ok and gets merged, I will try and see if I can create a quick pull request on the workflow for evac and synb0 as well.

**Comment by pjsjongsung** [comment]
@skoudoro sorry it took some time. Had some issues with tests. Also changed so that we can prevent diffusion images as input. Should be good to go!

**Review by skoudoro — APPROVED** [review]
All good, I played a bit with this method and it is great!

I will go ahead and merge this. Thank you @pjsjongsung!

**Pull request merged**